README
= Py Suite (PySlices, PyCrust, etc.) =

The Py Suite in wxPython, originally written by Patrick O'Brien, was and is an extremely popular programming shell for the Python programming language.

Old website: http://www.wxpython.org/py.php

It has been used by multiple projects (Stani's Python Editor, many Enthought tools, etc...) and supports calltips, history storage, code completion, and many other features in a solid, GUI based python shell.

I decided in 2008 after using Pycrust to program Python for a couple of years that I wanted to modify the shell to suit my needs.  I had been used to working with Mathematica, so I adore a notebook-style interface where you can write multiple lines of code that can be run and rerun in chunks.  After using the following "if 1:" construct in PyCrust for over a year:

{{{
if 1:
    <Many lines indented code...>
}}}

I finally decided to modify the shell myself to support real code blocks, and PySlices was born into the Py family...  It supports input and output "slices" (grouped lines of code / output text), re-running of old commands IN PLACE, multiple commands per "slice", and even saving and loading the .pyslices format (which can be renamed .py and used directly as python code)!

Also notable is that these patches add some often-coveted functionality to BOTH PyCrust and PySlices -- support for ipython-style magic features!

Typing a space for functions:
  "f 1" will auto-convert to "f(1)"


Support for unix-style directory commands:
  "cd", "ls", and "pwd" all work as expected


Support for quick shell commands with the ! mark:
  "!echo test" will print "test" in the shell


There is a lot more to say, but I've said enough... try it out, and read the tutorials!

If you want to install, you will need python as well as wxPython 2.8 or later.


----
Also, if you notice, there is also an experimental shell called SymPySlices.

SymPySlices REQUIRES *sympy*, and is still somewhat experimental.  It also uses the ast module and REQUIRES *Python 2.6 or later*.

That said, it is really cool!

It is basically a hack to allow the use of unicode characters AS python code

It also has a setting to allow automatic symbol creation from unknown variables.

so, I could write:

{{{
∫(x²,x)
}}}

and the result would be `x**3/3`  !

The entry method is to press the escape character and type a shortcuts phrase, ie:

{{{
<ESC>theta<ESC>
}}}

yeilds θ !

A large number of latex names are already supported, but every unicode character is supported somehow (see symbolConversionTable.py for a complete listing of names).

Files are actually saved in a plain text (ascii) format to avoid conversion issues.

SymPySlices also has some (slightly flaky) support for infix operators...

So, I could define a dot product operator to allow me to do the following:

{{{
[1,2,3,4]⋅[4,3,2,1]
}}}

yeilds 20.

Note that intergal and dot product are actually pre-defined using SymPySlicesDefaults.py, now included in wxpysuite.  These behaviors can be easily overridden using:

{{{
∫ = newFunction
}}}

and:

{{{
__DotOperator__ = newFunction
}}}

There are a lot more symbols defined, so dig in and check them out!

My basic hope for this aspect of the project is to spread some ideas about how to use unicode and python together with sympy to try to create computer algebra system that is a little bit simpler than SAGE.


----
== Installing and Running ==

You can either install from here or from PyPI (package name is wx_py).

To install this version, unpack the tar.gz.

At the command line, go to the new directory and type:

{{{
python setup.py install
}}}

on linux and mac, you may need to type:

{{{
sudo python setup.py install
}}}

Run PyCrust with this command:

{{{
python -m wx_py.PyCrust
}}}

Run PySlices with this command:

{{{
python -m wx_py.PySlices
}}}

Run SymPySlices with this command:

{{{
python -m wx_py.SymPySlices
}}}

To see a sample configuration file for SymPySlices, you may want to look at SymPySlicesDefaults.py (in the Downloads section).  If you copy these contents into the startup script (Options Menu -> Startup -> Edit Startup Script) then SymPySlices' operators will behave the way I've described above.  I'm still working out what the true default behavior should be and I want people to be able to customize this, so I have left this out of the installed version intentionally.


----
== Screenshots ==

http://wxpysuite.googlecode.com/svn/trunk/PyCrustScreenshot.png

http://wxpysuite.googlecode.com/svn/trunk/PySlicesScreenshot.png


----
=== Other Projects ===

Just wanted to note that there are other cool projects out there, too.

IPython, SAGE, Symbide, and Reinteract are some of the best!  SAGE is an especially promising project!


----
BTW, I have created the module with the name wx_py so it will be trivial to convert code to wx.py if and when you update to a version included with wxPython itself (not that you should even use Py Suite code in your programming...)

Also, a legality: I listed this a LGPL, but it is really under the wxWidgets licence (essentially LGPL)

= ... =



CHANGES.txt
0.9.8.12 (4/12/2015)
--------------------
Work in progress towards fixing wxPython 3.0 compatibility problems;
 this is nominally working, but still has rough edges.
A lot of internal refactoring (stop abusing keywords, etc.)

0.9.8.11 (4/30/2013)
--------------------
Minor bugfix related to strings like '\''
Minor bugfix for sliceshell.py
Lots of beind-the-scenes changes with move to git version control.
Source is now hosted at github and bitbucket (and google code and gitorious)

0.9.8.10 (3/27/2012)
-------------------
Fixed a bug when loading SymPySlices Files

0.9.8.9 (3/13/2012)
-------------------
Updated the .ico files and added postinstall script to build windows
Desktop Shortcuts. First build of exe installers.

0.9.8.3 - 0.9.8.8) (1/17/2012)
-------------------
Added a comment you can change to reveal a blank screen on startup.
Made svg icons and remade the pngs.
Made the shell notify the OS every time a command is finished (helpful
with long-running processes so you can check them visually from the os).

0.9.8.7 (6/23/2011)
-------------------
Added a couple more items to the symbol table for SymPySlices
Project seems to have stabilized; no new bugs found.

0.9.8.6 (1/17/2011)
-------------------
In SymPySlices, changed writen unicode characters from utf-8 directly to unicode
because of a bug in Windows version
Affected variables and functions:
* ESC_SYMBOL (DictWriter and symbolConversion)
* FromName (Dicts)
* Ascii2Unicode
* FormatAsciiForDisplay
* CheckForAutoSymbolReplacement

0.9.8.5 - 0.9.8.5.2 (1/16/2011)
-------------------
Changed magic to auto-transform strings to ''' ''' instead of " ".
The quotes were conflicting with path name quotes for shell commands...

0.9.8.4.7 - 0.9.8.5 (7/14/2010)
-------------------
Created custom Save Dialog to use the good "Human Interface" practice of
giving specific verbs to buttons (Don't Save, Save, Cancel)...

Fixed a bug in frame.py that kept the Hide Folding Markers option from displaying.

Partially fixed a bug with readline. (Still problems if you delete stuff
printed by write, not written by user).

0.9.8.4.6 (6/2/2010)
-------------------
Merged cvs tags with new changes from wxPython.

Fixed some issues with SymPySlicesDefaults (made it so numpy is optional).
Before, numpy would be required or else errors would pop up...

Tweaked magic to play nicely with symbols (especially operators).

0.9.8.4.5 (5/31/2010)
-------------------
Fixed some tutorial typos:
  Changed Description of shell mode...
  In sympyslices, 'not' is now 'now'
  Change dot variables to c,d

Added cphi, ctheta, etc... to the list of supported names...

Changed to allow a second defaults file specifically for overloading special characters!
This file is actually a module within wx.py that contains one large string called
"commands" to be run by the interpreter line by line...

SymPySlices Tutorial has really been tuned up to account for these and other changes...

0.9.8.4 - 0.9.8.4.4 (5/31/2010)
-------------------
Fixed some issues with file save an load on windows.
1. Files were not loading properly at all
2. Files were being saved with '\r\r\n' endings instead of '\r\n'

Fixed tutorial to be smarter about placement given different prior print-outs.

0.9.8.3.6 (5/27/2010)
-------------------

Added more symbols (notably '~~' and '->') to sympyslices table.

0.9.8.3.5 (4/11/2010)
-------------------
Fixed bug that led to initial current directory always being '~' (user folder).
Now, remains in directory launched from instead.

0.9.8.3.1 - 3 (2/8/2010)
-------------------
Added "Save A Copy" menu item.
Cleaned up the "Save As" code to properly handle file overwrite, etc...

Continued to improve the symbolConversionTable.

Added FindPrevious Menu option, changed FindNext keybinding to Ctrl-G
Made PySlices and SympySlices have a find-replace dialog instead
   (still need to implement the replace features, though...)

Fixed a bug in the "New" menu... made it revert to a blank shell
   (which is the best thing to do without getting into tab/window management)

0.9.8.3 (1/22/2010)
-------------------
Major update to the symbolConversion schema for auto-conversion of
unicode in SymPySlices.

0.9.8.2.6 - 2.8 (1/13/2010)
-------------------
Made the shell mode a lot better by preempting the interpreter using parse.py
Now, rather than the two-return system, it mimics the regular python interpreter!
Also, a custom-printed SyntaxError message will print if the syntax is not valid...
And to help the user out,the cursor will return to the line where the error was found!!

0.9.8.2.5 (1/10/2010)
-------------------
Added support for remebering last opening directory.  This seems to be a
more standard feel.

0.9.8.2.3 / 2.4 (12/21/2009)
-------------------

Added support for smarter parsing... now:
if 1: # Comment
supports auto-indent

Separated out the parsing function (testForContinuations) into parse.py
It now checks for string, line, parenthetical, and indentation-based continuations...

The biggest change in all this, however, is that the shell mode acts much more like a
traditional python interpreter... no more double-returns... :)


0.9.8.2 / 2.1 / 2.2 (12/18/2009)
-------------------
Added nominal readline support (fun readline character disappears after running...
                                SO Much simpler this way...)
Added safety to prevent running readline input from a different input slice...

Fixed a bug in error printing (was no background color)

Fixed a bug in OnChar (shouldn't have been calling chr() on things like Alt-UpArrow)

0.9.8.1.17 (current)
-------------------
<current>

0.9.8.1.16 (12/10/2009) ??
-------------------
Removed most unnecessary print commands
Fixed another bug with triple quotes:
"""

"""
print 1

Did not run the 'print 1' command...
Also, this now runs properly:
1 #+ \
2

0.9.8.1.15 (12/2/09)
-------------------
Fixed a bug in magic that allowed it to deal with commented triple quotes:

# """
ls

Will now respond correctly...

crust.py
"""Crust combines the shell and filling into one control."""

__author__ = "Patrick K. O'Brien <pobrien@orbtech.com>"
__cvsid__ = "$Id: crust.py 44235 2007-01-17 23:05:14Z RD $"
__revision__ = "$Revision: 44235 $"[11:-2]

import wx

import os
import pprint
import re
import sys

import dispatcher
import editwindow
from filling import Filling
import frame
from shell import Shell
from version import VERSION


class Crust(wx.SplitterWindow):
    """Crust based on SplitterWindow."""

    name = 'Crust'
    revision = __revision__
    sashoffset = 300

    def __init__(self, parent, id=-1, pos=wx.DefaultPosition,
                 size=wx.DefaultSize, style=wx.SP_3D|wx.SP_LIVE_UPDATE,
                 name='Crust Window', rootObject=None, rootLabel=None,
                 rootIsNamespace=True, intro='', locals=None,
                 InterpClass=None,
                 startupScript=None, execStartupScript=True,
                 *args, **kwds):
        """Create Crust instance."""
        wx.SplitterWindow.__init__(self, parent, id, pos, size, style, name)

        # Turn off the tab-traversal style that is automatically
        # turned on by wx.SplitterWindow.  We do this because on
        # Windows the event for Ctrl-Enter is stolen and used as a
        # navigation key, but the Shell window uses it to insert lines.
        style = self.GetWindowStyle()
        self.SetWindowStyle(style & ~wx.TAB_TRAVERSAL)
        
        self.shell = Shell(parent=self, introText=intro,
                           locals=locals, InterpClass=InterpClass,
                           startupScript=startupScript,
                           execStartupScript=execStartupScript,
                           *args, **kwds)
        
        self.editor = self.shell
        if rootObject is None:
            rootObject = self.shell.interp.locals
        self.notebook = wx.Notebook(parent=self, id=-1)
        self.shell.interp.locals['notebook'] = self.notebook
        self.filling = Filling(parent=self.notebook,
                               rootObject=rootObject,
                               rootLabel=rootLabel,
                               rootIsNamespace=rootIsNamespace)
        # Add 'filling' to the interpreter's locals.
        self.shell.interp.locals['filling'] = self.filling
        self.notebook.AddPage(page=self.filling, text='Namespace', select=True)
        
        self.display = Display(parent=self.notebook)
        self.notebook.AddPage(page=self.display, text='Display')
        # Add 'pp' (pretty print) to the interpreter's locals.
        self.shell.interp.locals['pp'] = self.display.setItem
        self.display.nbTab = self.notebook.GetPageCount()-1
        
        self.calltip = Calltip(parent=self.notebook)
        self.notebook.AddPage(page=self.calltip, text='Calltip')
        
        self.sessionlisting = SessionListing(parent=self.notebook)
        self.notebook.AddPage(page=self.sessionlisting, text='History')
        
        self.dispatcherlisting = DispatcherListing(parent=self.notebook)
        self.notebook.AddPage(page=self.dispatcherlisting, text='Dispatcher')

        
        # Initialize in an unsplit mode, and check later after loading
        # settings if we should split or not.
        self.shell.Hide()
        self.notebook.Hide()
        self.Initialize(self.shell)
        self._shouldsplit = True
        wx.CallAfter(self._CheckShouldSplit)
        self.SetMinimumPaneSize(100)

        self.Bind(wx.EVT_SIZE, self.SplitterOnSize)
        self.Bind(wx.EVT_SPLITTER_SASH_POS_CHANGED, self.OnChanged)
        self.Bind(wx.EVT_SPLITTER_DCLICK, self.OnSashDClick)

    def _CheckShouldSplit(self):
        if self._shouldsplit:
            self.SplitHorizontally(self.shell, self.notebook, -self.sashoffset)
            self.lastsashpos = self.GetSashPosition()
        else:
            self.lastsashpos = -1
        self.issplit = self.IsSplit()       

    def ToggleTools(self):
        """Toggle the display of the filling and other tools"""
        if self.issplit:
            self.Unsplit()
        else:
            self.SplitHorizontally(self.shell, self.notebook, -self.sashoffset)
            self.lastsashpos = self.GetSashPosition()
        self.issplit = self.IsSplit()

    def ToolsShown(self):
        return self.issplit

    def OnChanged(self, event):
        """update sash offset from the bottom of the window"""
        self.sashoffset = self.GetSize().height - event.GetSashPosition()
        self.lastsashpos = event.GetSashPosition()
        event.Skip()

    def OnSashDClick(self, event):
        self.Unsplit()
        self.issplit = False

    # Make the splitter expand the top window when resized
    def SplitterOnSize(self, event):
        splitter = event.GetEventObject()
        sz = splitter.GetSize()
        splitter.SetSashPosition(sz.height - self.sashoffset, True)
        event.Skip()


    def LoadSettings(self, config):
        self.shell.LoadSettings(config)
        self.filling.LoadSettings(config)

        pos = config.ReadInt('Sash/CrustPos', 400)
        wx.CallAfter(self.SetSashPosition, pos)
        def _updateSashPosValue():
            sz = self.GetSize()
            self.sashoffset = sz.height - self.GetSashPosition()
        wx.CallAfter(_updateSashPosValue)
        zoom = config.ReadInt('View/Zoom/Display', -99)
        if zoom != -99:
            self.display.SetZoom(zoom)
        self.issplit = config.ReadInt('Sash/IsSplit', True)
        if not self.issplit:
            self._shouldsplit = False

    def SaveSettings(self, config):
        self.shell.SaveSettings(config)
        self.filling.SaveSettings(config)

        if self.lastsashpos != -1:
            config.WriteInt('Sash/CrustPos', self.lastsashpos)
        config.WriteInt('Sash/IsSplit', self.issplit)
        config.WriteInt('View/Zoom/Display', self.display.GetZoom())
        
class Display(editwindow.EditWindow):
    """STC used to display an object using Pretty Print."""

    def __init__(self, parent, id=-1, pos=wx.DefaultPosition,
                 size=wx.DefaultSize,
                 style=wx.CLIP_CHILDREN | wx.SUNKEN_BORDER,
                 static=False):
        """Create Display instance."""
        editwindow.EditWindow.__init__(self, parent, id, pos, size, style)
        # Configure various defaults and user preferences.
        self.SetReadOnly(True)
        self.SetWrapMode(False)
        if not static:
            dispatcher.connect(receiver=self.push, signal='Interpreter.push')

    def push(self, command, more):
        """Receiver for Interpreter.push signal."""
        self.Refresh()

    def Refresh(self):
        if not hasattr(self, "item"):
            return
        self.SetReadOnly(False)
        text = pprint.pformat(self.item)
        self.SetText(text)
        self.SetReadOnly(True)

    def setItem(self, item):
        """Set item to pretty print in the notebook Display tab."""
        self.item = item
        self.Refresh()
        if self.GetParent().GetSelection() != self.nbTab:
            focus = wx.Window.FindFocus()
            self.GetParent().SetSelection(self.nbTab)
            wx.CallAfter(focus.SetFocus)
            

# TODO: Switch this to a editwindow.EditWindow
class Calltip(wx.TextCtrl):
    """Text control containing the most recent shell calltip."""

    def __init__(self, parent=None, id=-1,ShellClassName='Shell'):
        style = (wx.TE_MULTILINE | wx.TE_READONLY | wx.TE_RICH2)
        wx.TextCtrl.__init__(self, parent, id, style=style)
        self.SetBackgroundColour(wx.Colour(255, 255, 208))
        self.ShellClassName=ShellClassName
        dispatcher.connect(receiver=self.display, signal=self.ShellClassName+'.calltip')

        df = self.GetFont()
        font = wx.Font(df.GetPointSize(), wx.TELETYPE, wx.NORMAL, wx.NORMAL)
        self.SetFont(font)

    def display(self, calltip):
        """Receiver for """+self.ShellClassName+""".calltip signal."""
        ## self.SetValue(calltip)  # Caused refresh problem on Windows.
        self.Clear()
        self.AppendText(calltip)


# TODO: Switch this to a editwindow.EditWindow
class SessionListing(wx.TextCtrl):
    """Text control containing all commands for session."""

    def __init__(self, parent=None, id=-1,ShellClassName='Shell'):
        style = (wx.TE_MULTILINE | wx.TE_READONLY |
                 wx.TE_RICH2 | wx.TE_DONTWRAP)
        wx.TextCtrl.__init__(self, parent, id, style=style)
        dispatcher.connect(receiver=self.addHistory,
                           signal=ShellClassName+".addHistory")
        dispatcher.connect(receiver=self.clearHistory,
                           signal=ShellClassName+".clearHistory")
        dispatcher.connect(receiver=self.loadHistory,
                           signal=ShellClassName+".loadHistory")

        df = self.GetFont()
        font = wx.Font(df.GetPointSize(), wx.TELETYPE, wx.NORMAL, wx.NORMAL)
        self.SetFont(font)

    def loadHistory(self, history):
        # preload the existing history, if any
        hist = history[:]
        hist.reverse()
        self.SetValue('\n'.join(hist) + '\n')
        self.SetInsertionPointEnd()

    def addHistory(self, command):
        if command:
            self.SetInsertionPointEnd()
            self.AppendText(command + '\n')

    def clearHistory(self):
        self.SetValue("")


class DispatcherListing(wx.TextCtrl):
    """Text control containing all dispatches for session."""

    def __init__(self, parent=None, id=-1):
        style = (wx.TE_MULTILINE | wx.TE_READONLY |
                 wx.TE_RICH2 | wx.TE_DONTWRAP)
        wx.TextCtrl.__init__(self, parent, id, style=style)
        dispatcher.connect(receiver=self.spy)

        df = self.GetFont()
        font = wx.Font(df.GetPointSize(), wx.TELETYPE, wx.NORMAL, wx.NORMAL)
        self.SetFont(font)

    def spy(self, signal, sender):
        """Receiver for Any signal from Any sender."""
        text = '%r from %s' % (signal, sender)
        self.SetInsertionPointEnd()
        start, end = self.GetSelection()
        if start != end:
            self.SetSelection(0, 0)
        self.AppendText(text + '\n')



class CrustFrame(frame.Frame, frame.ShellFrameMixin):
    """Frame containing all the PyCrust components."""

    name = 'CrustFrame'
    revision = __revision__


    def __init__(self, parent=None, id=-1, title='PyCrust',
                 pos=wx.DefaultPosition, size=wx.DefaultSize,
                 style=wx.DEFAULT_FRAME_STYLE,
                 rootObject=None, rootLabel=None, rootIsNamespace=True,
                 locals=None, InterpClass=None,
                 config=None, dataDir=None,
                 *args, **kwds):
        """Create CrustFrame instance."""
        frame.Frame.__init__(self, parent, id, title, pos, size, style,
                             shellName='PyCrust')
        frame.ShellFrameMixin.__init__(self, config, dataDir)
        
        if size == wx.DefaultSize:
            self.SetSize((800, 600))
        
        intro = 'PyCrust %s - The Flakiest Python Shell' % VERSION
        
        self.SetStatusText(intro.replace('\n', ', '))
        self.crust = Crust(parent=self, intro=intro,
                           rootObject=rootObject,
                           rootLabel=rootLabel,
                           rootIsNamespace=rootIsNamespace,
                           locals=locals,
                           InterpClass=InterpClass,
                           startupScript=self.startupScript,
                           execStartupScript=self.execStartupScript,
                           *args, **kwds)
        self.shell = self.crust.shell

        # Override the filling so that status messages go to the status bar.
        self.crust.filling.tree.setStatusText = self.SetStatusText
        
        # Override the shell so that status messages go to the status bar.
        self.shell.setStatusText = self.SetStatusText
        
        self.shell.SetFocus()
        self.LoadSettings()


    def OnClose(self, event):
        """Event handler for closing."""
        self.SaveSettings()
        self.crust.shell.destroy()
        self.Destroy()


    def OnAbout(self, event):
        """Display an About window."""
        title = 'About PyCrust'
        text = 'PyCrust %s\n\n' % VERSION + \
               'Yet another Python shell, only flakier.\n\n' + \
               'Half-baked by Patrick K. O\'Brien,\n' + \
               'the other half is still in the oven.\n\n' + \
               'Shell Revision: %s\n' % self.shell.revision + \
               'Interpreter Revision: %s\n\n' % self.shell.interp.revision + \
               'Platform: %s\n' % sys.platform + \
               'Python Version: %s\n' % sys.version.split()[0] + \
               'wxPython Version: %s\n' % wx.VERSION_STRING + \
               ('\t(%s)\n' % ", ".join(wx.PlatformInfo[1:]))
        dialog = wx.MessageDialog(self, text, title,
                                  wx.OK | wx.ICON_INFORMATION)
        dialog.ShowModal()
        dialog.Destroy()


    def ToggleTools(self):
        """Toggle the display of the filling and other tools"""
        return self.crust.ToggleTools()

    def ToolsShown(self):
        return self.crust.ToolsShown()

    def OnHelp(self, event):
        """Show a help dialog."""
        frame.ShellFrameMixin.OnHelp(self, event)

    def LoadSettings(self):
        if self.config is not None:
            frame.ShellFrameMixin.LoadSettings(self)
            frame.Frame.LoadSettings(self, self.config)
            self.crust.LoadSettings(self.config)


    def SaveSettings(self, force=False):
        if self.config is not None:
            frame.ShellFrameMixin.SaveSettings(self,force)
            if self.autoSaveSettings or force:
                frame.Frame.SaveSettings(self, self.config)
                self.crust.SaveSettings(self.config)


    def DoSaveSettings(self):
        if self.config is not None:
            self.SaveSettings(force=True)
            self.config.Flush()
    
